extension/src/frequency.rs

```diff
@@ -403,100 +400,212 @@ pub mod toolkit_experimental {
+                None => return state,
+                Some(value) => value,
+            };
+            let mut state = match state {
+                None => {
+                    let typ = value.oid();
+                    let collation = if fcinfo.is_null() {
+                        Some(100) // TODO: default OID, there should be a constant for this
+                    } else {
+                        get_collation(fcinfo)
+                    };
+                    SpaceSavingTransState::topn_agg_from_type_id(skew, n as u32, typ, collation).into()
+                },
+                Some(state) => state,
+            };
+    
+            state.add(value.into());
+            Some(state)
+        })
     }
-
-    ron_inout_funcs!(FrequencyAggregate);
 }
 
-#[aggregate]
-impl toolkit_experimental::freq_agg {
-    type State = FrequencyTransState;
-
-    const PARALLEL_SAFE: bool = true;
-
-    fn transition(
-        state: Option<State>,
-        #[sql_type("double precision")] freq: f64,
-        #[sql_type("AnyElement")] value: Option<AnyElement>,
-        fcinfo: pg_sys::FunctionCallInfo,
-    ) -> Option<State> {
-        let value = match value {
-            None => return state,
-            Some(value) => value,
-        };
-        let mut state = match state {
-            None => unsafe {
-                let typ = value.oid();
-                let collation = if fcinfo.is_null() {
-                    Some(100) // TODO: default OID, there should be a constant for this
-                } else {
-                    get_collation(fcinfo)
-                };
-                FrequencyTransState::from_type_id(freq, typ, collation)
-            },
-            Some(state) => state,
-        };
-
-        state.add(value.into());
-        Some(state)
+#[pg_extern(schema = "toolkit_experimental", immutable, parallel_safe)]
+pub fn freq_agg_trans(
+    state: Internal,
+    freq: f64,
+    value: Option<AnyElement>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Internal {
+    freq_agg_trans_inner(unsafe{ state.to_inner() }, freq, value, fcinfo).internal()
+}
+pub fn freq_agg_trans_inner(
+    state: Option<Inner<SpaceSavingTransState>>,
+    freq: f64,
+    value: Option<AnyElement>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Inner<SpaceSavingTransState>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            let value = match value {
+                None => return state,
+                Some(value) => value,
+            };
+            let mut state = match state {
+                None => {
+                    let typ = value.oid();
+                    let collation = if fcinfo.is_null() {
+                        Some(100) // TODO: default OID, there should be a constant for this
+                    } else {
+                        get_collation(fcinfo)
+                    };
+                    SpaceSavingTransState::freq_agg_from_type_id(freq, typ, collation).into()
```

Comment:
Looks like this line is the only difference between this function and `topn_agg_trans_inner` - we could pass `|typ, collation| SpaceSavingTransState::freq_agg_from_type_id(freq, typ, collation)` to a factored-out transition function.